### ISSUE TITLE
SP9: Camera support

### ISSUE THREAD

**Issue by peter-marshall5** [linux-surface/linux-surface#1064]
I would like to help to get the cameras supported on my device.  It is listed under `IPU4` for the ISP, so I understand that camera support may be a long way away.

# Environment
- Hardware Model: Surface Pro 9 (Intel)
- Kernel version: 6.1.12
- Distribution: Fedora Silverblue

I currently do not have a `dmesg` log but will add it ASAP.

**Comment by peter-marshall5** [comment]
From my research, it seems like the SP9 uses the `IPU6` ISP that is found in the Intel Alder Lake CPU. There's a partially-open source driver made by Intel to get the cameras working on Linux [here](https://github.com/intel/ipu6-drivers).

**Comment by BloodyIron** [comment]
+1

**Comment by derkrasseleo** [comment]
I am also using a SP9 with arch, so if I can do some testing lmk

**Comment by yaoengine** [comment]
So is there any progress now? I am using SP8 and seems it has the same camera module with SP9, hope can be help to get it working.

**Comment by BloodyIron** [comment]
Yeah I'm still interested in the camera working for SP9.

**Comment by StollD** [comment]
This really depends on Intel and their willingness to upstream the drivers for IPU6 and to work with distributions on integrating it. Unfortunately, these days they seem to take the approach of throwing their code over the wall and then running away from it as fast as possible.

There are ways to make it work with out of tree drivers and a daemon that copies the camera data into a v4l2loopback device, and some distributions package that, e.g. https://hansdegoede.livejournal.com/27276.html for Fedora.

But I dont remember if anyone has tried that yet, so I dont know if it will work on surfaces.

**Comment by yaoengine** [comment]
I used the code in `https://github.com/intel/ipu6-drivers/tree/iotg_ipu6`, with linux-surface kernel patch, the modules can be complied and loaded.
But seems camera still not work.
The part of dmsg is like this:

```
[    2.873921] EDAC igen6: v2.5.1
[    2.881265] intel-ipu6 0000:00:05.0: enabling device (0000 -> 0002)
[    2.881487] intel-ipu6 0000:00:05.0: Device 0x9a19 (rev: 0x1)
[    2.881537] intel-ipu6 0000:00:05.0: physical base address 0x603c000000
[    2.881538] intel-ipu6 0000:00:05.0: mapped as: 0x000000009be5056a
[    2.884001] intel-ipu6 0000:00:05.0: IPU in secure mode 
[    2.884003] intel-ipu6 0000:00:05.0: IPU secure touch = 0x0
[    2.884005] intel-ipu6 0000:00:05.0: IPU camera mask = 0xff
[    2.895409] int3472-discrete INT3472:01: GPIO type 0x08 unknown; the sensor may not work
[    2.895725] intel-ipu6 0000:00:05.0: IPC reset done 
[    2.895727] intel-ipu6 0000:00:05.0: cpd file name: intel/ipu6_fw.bin
[    2.898607] intel-ipu6 0000:00:05.0: FW version: 20201222
[    2.899915] intel-ipu6 0000:00:05.0: Found supported sensor INT33BE:00
[    2.900146] intel-ipu6 0000:00:05.0: Connected 1 cameras
[    2.902867] intel-ipu6 0000:00:05.0: Sending BOOT_LOAD to CSE
[    2.912196] intel-ipu6 0000:00:05.0: Sending AUTHENTICATE_RUN to CSE
[    2.920323] int3472-discrete INT3472:02: unknown \_SB.GPI0 pin number mismatch _DSM 15 resource 10
[    2.920326] int3472-discrete INT3472:02: GPIO type 0x10 unknown; the sensor may not work
```
As I am not familiar with linux kernel, so I donot know how to going on.
Anyone want to go on with it?

**Comment by yaoengine** [comment]
it seems `drivers/platform/x86/intel/int3472/discrete.c` only process the following GPIO pin:

```C++
 * function of the GPIO pin:
 *
 * 0x00 Reset
 * 0x01 Power down
 * 0x0b Power enable
 * 0x0c Clock enable
 * 0x0d Privacy LED
```
But it get 0x08, and report the error:
`[    2.895409] int3472-discrete INT3472:01: GPIO type 0x08 unknown; the sensor may not work`

**Comment by BloodyIron** [comment]
Any news on this? :)

**Comment by derkrasseleo** [comment]
Maybe this is useful: https://github.com/stefanpartheym/archlinux-ipu6-webcam

**Comment by BloodyIron** [comment]
Has anyone tested that on an SP9?

**Comment by derkrasseleo** [comment]
Yes, this is what I got: https://github.com/linux-surface/linux-surface/issues/91#issuecomment-1880154137

**Comment by daegalus** [comment]
https://www.phoronix.com/news/Intel-IPU6-Firmware-Git

**Comment by BloodyIron** [comment]
So what exactly is the state for camera on sp9 at this point? I thought Linux kernel versions from this year addressed it, but uhh so far I'm not seeing it working on my end.

**Comment by derkrasseleo** [comment]
I am also unsure what the state is at this point. With `6.12.7-arch1-1-surface` I get multiple (many) instances of `ipu6` and one "virtual camera" in both cheese and WebRTC, but none of them seem to have any other output than a black screen. Does anyone know why there are so many of them and if there a way to clean those up?
![image](https://github.com/user-attachments/assets/0f2f2038-9e17-4efc-a0a2-15a25096ab1d)

**Comment by BloodyIron** [comment]
I'm rocking 6.12.3-surface-2 on Ubuntu 24.04 on a Surface Pro 9, and I CANNOT see the camera device at all...

**Comment by BloodyIron** [comment]
I just remembered I had blacklisted (/etc/modprobe.d/blacklist.conf" the "intel-ipu6" and "intel-ipu6-isys" modules back in Sept of 2024 due to boot problems.

I uncommented them, and now I'm getting problems in dmesg I'm pretty sure relating to those modules crashing or something. Still reading through dmesg but after un-blacklisting them and rebooting, still no camera.

**Comment by BloodyIron** [comment]
I'm on "BIOS 9.20.143 02/01/2023" by the way as reported by dmesg, but last I checked you only can update BIOS versions via Windows for the Surface Pro 9, and not sure if that changes anything for this anyways.

**Comment by BloodyIron** [comment]
Yeah I can't get this damn thing working. I'm probably just going to put Windows back on this and sell it. Appreciate the community working to make Linux run on these devices, but after a long while of using it I kinda want to just switch to a Ryzen laptop. More repairable, and this being a tablet isn't really giving me much over a laptop. Plus a laptop sits way better on an actual lap.

I had expected intel to do a proper job on this driver, looks like they haven't. No wonder they're in trouble.

**Comment by jetm** [comment]
@BloodyIron If you don't mind, you can try Arch Linux and https://github.com/stefanpartheym/archlinux-ipu6-webcam. It works fine on my laptop with the same camera.

**Comment by derkrasseleo** [comment]
> @BloodyIron If you don't mind, you can try Arch Linux and https://github.com/stefanpartheym/archlinux-ipu6-webcam. It works fine on my laptop with the same camera.

Interesting, I am running arch and tried that and it doesn't work on my SP9. Did you install anything else?

**Comment by jetm** [comment]
Hi @derkrasseleo
 
> Interesting, I am running arch and tried that and it doesn't work on my SP9. Did you install anything else?

Just make sure to use the kernel LTS `6.6.y`. It doesn't work with newer kernels.

**Comment by letanphuc** [comment]
Any update on this?

**Comment by petm5** [comment]
My SP9 uses an alternate ACPI device name for the front camera that isn't listed in the `ov5693` driver. A simple kernel patch attaached below makes the driver probe successfully, and qcam makes the camera light illuminate, but something still seems to be missing as I'm not getting any video. I haven't had any luck with the rear camera, the `ov13858` driver fails with error -22.

[sp9-front-camera-acpi-id.patch](https://github.com/user-attachments/files/22854868/sp9-front-camera-acpi-id.patch)

These dmesg lines seem relevant:
```
[    8.622467] int3472-discrete INT3472:00: GPIO type 0x08 unknown; the sensor may not work
[    8.649991] int3472-discrete INT3472:01: GPIO type 0x10 unknown; the sensor may not work
[52049.517603] ov5693 i2c-OVTI5693:00: supply dovdd not found, using dummy regulator
[52049.517691] ov5693 i2c-OVTI5693:00: supply dvdd not found, using dummy regulator
```

**Comment by qzed** [comment]
Apparently @toorajtaraz got it working: https://github.com/linux-surface/linux-surface/issues/1865#issuecomment-3391583970 🎉

**Comment by amidukr** [comment]
> My SP9 uses an alternate ACPI device name for the front camera that isn't listed in the `ov5693` driver. A simple kernel patch attaached below makes the driver probe successfully, and qcam makes the camera light illuminate, but something still seems to be missing as I'm not getting any video. I haven't had any luck with the rear camera, the `ov13858` driver fails with error -22.
> 
> [sp9-front-camera-acpi-id.patch](https://github.com/user-attachments/files/22854868/sp9-front-camera-acpi-id.patch)
> 
> These dmesg lines seem relevant:
> 
> ```
> [    8.622467] int3472-discrete INT3472:00: GPIO type 0x08 unknown; the sensor may not work
> [    8.649991] int3472-discrete INT3472:01: GPIO type 0x10 unknown; the sensor may not work
> [52049.517603] ov5693 i2c-OVTI5693:00: supply dovdd not found, using dummy regulator
> [52049.517691] ov5693 i2c-OVTI5693:00: supply dvdd not found, using dummy regulator
> ```

Same issue with Surface Pro 9 on Ubuntu with latest surface kernel.

However I've found more elegant way to solve same issue without patching. Modprobe when load modules it can use alias to bind to device with no standard names. So what i've did, I've added this file:

```bash
> $ cat /etc/modprobe.d/ov5693.conf
alias acpi*:OVTI5693:* ov5693
alias i2c:ov5693 ov5693
```

Before I add this line, linux kernel was not even trying to load ov5693 module. I was able to observe this by looking into:

```bash
ls ls /sys/bus/i2c/drivers
```

When I've added `/etc/modprobe.d/ov5693.conf`, output for lsmod also become different:

```bash
lsmod | grep -i -e ov -e 5693

ov5693                 24576  0
ov13858                24576  0
v4l2_fwnode            36864  3 ov5693,intel_ipu6_isys,ov13858
v4l2_async             28672  4 v4l2_fwnode,ov5693,intel_ipu6_isys,ov13858
videodev              344064  7 v4l2_async,v4l2_fwnode,videobuf2_v4l2,ov5693,v4l2loopback,intel_ipu6_isys,ov13858
mc                     81920  7 v4l2_async,videodev,videobuf2_v4l2,ov5693,intel_ipu6_isys,ov13858,videobuf2_common
```

Obviously module v4l2loopback-dkms is required for `ov13858` module to work.

The way I've figure out to put file `/etc/modprobe.d/ov5693.conf`, linux kernel on Ubuntu for Surface Pro 9 was able to figure back camera `ov13858`, so I've tried to add similar setting for `ov5693`.

Actually there are few device file, that linux providing by scanning hardware:
```bash
cat /sys/bus/acpi/devices/OVTI5693:00/modalias
acpi:OVTI5693:OVTI5693:
```

```bash
cat /sys/bus/i2c/devices/i2c-OVTI5693:00/name
OVTI5693:00
```


This two errors, what it looks to me:
- `GPIO type 0x08 unknown;` and `GPIO type 0x10 unknown;`
ov5693 tried to interact with a device, but device even that it has marking of OVTI5693 it still following some who knows what proprietary protocol.
